pinot-server/src/main/java/org/apache/pinot/server/starter/helix/OffsetBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,119 @@
+
+  // helper variable
+  private final Set<String> _caughtUpSegments = new HashSet<>();
+
+  public OffsetBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager, Set<String> consumingSegments) {
+    _instanceDataManager = instanceDataManager;
+    _consumingSegments = consumingSegments;
+  }
+
+  public boolean haveAllConsumingSegmentsReachedStreamLatestOffset() {
+    boolean allSegsReachedLatest = true;
+    for (String segName : _consumingSegments) {
+      if (_caughtUpSegments.contains(segName)) {
+        continue;
+      }
+      TableDataManager tableDataManager = getTableDataManager(segName);
+      if (tableDataManager == null) {
+        LOGGER.info("TableDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      SegmentDataManager segmentDataManager = tableDataManager.acquireSegment(segName);
+      if (segmentDataManager == null) {
+        LOGGER.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      if (!(segmentDataManager instanceof LLRealtimeSegmentDataManager)) {
+        // There's a possibility that a consuming segment has converted to a committed segment. If that's the case,
+        // segment data manager will not be of type LLRealtime.
+        LOGGER.info("Segment {} is already committed and is considered caught up.", segName);
+        _caughtUpSegments.add(segName);
+        releaseSegment(tableDataManager, segmentDataManager);
+        continue;
+      }
+      LLRealtimeSegmentDataManager rtSegmentDataManager = (LLRealtimeSegmentDataManager) segmentDataManager;
+      StreamPartitionMsgOffset latestIngestedOffset = rtSegmentDataManager.getCurrentOffset();
+      StreamPartitionMsgOffset latestStreamOffset = rtSegmentDataManager.getLatestStreamOffsetAtStartupTime();
+      releaseSegment(tableDataManager, segmentDataManager);
+      if (latestStreamOffset == null || latestIngestedOffset == null) {
+        LOGGER.info("Null offset found for segment {} - latest stream offset: {}, latest ingested offset: {}. "
+            + "Will check consumption status later", segName, latestStreamOffset, latestIngestedOffset);
+        return false;
+      }
+      if (latestIngestedOffset.compareTo(latestStreamOffset) < 0) {
+        LOGGER.info("Latest ingested offset {} in segment {} is smaller than stream latest available offset {} ",
+            latestIngestedOffset, segName, latestStreamOffset);
+        allSegsReachedLatest = false;
+        continue;
+      }
+      LOGGER.info("Segment {} with latest ingested offset {} has caught up to the latest stream offset {}", segName,
+            latestIngestedOffset, latestStreamOffset);
+      _caughtUpSegments.add(segName);
+    }
+    return allSegsReachedLatest;
+  }
+
+  void releaseSegment(TableDataManager tableDataManager, SegmentDataManager segmentDataManager) {
```

Comment:
why a separate method?

Reply:
Because it was used in two places. Now with the refactoring, it's only used once; hence no need for function.